File: ios/DisbandiOS/Services/Call/CallSignal.swift
import Foundation

/// A 1:1 call signaling message, wire-compatible with the desktop app's
/// `CallSignal` (sent as Realtime broadcasts on `call-user:<id>` and
/// `call:<callId>` channels, event "call").
struct CallSignal: Codable, Sendable {
    /// "ring" | "accept" | "reject" | "cancel" | "offer" | "answer" | "ice" | "leave"
    var type: String
    var from: String
    var to: String?
    var callId: String?
    var callerName: String?
    var rejecterName: String?
    var sdp: CallSdp?
    var candidate: CallIceCandidate?
}

/// Mirrors `RTCSessionDescriptionInit` JSON ({ type, sdp }).
struct CallSdp: Codable, Sendable {
    var type: String
    var sdp: String
}

/// Mirrors `RTCIceCandidateInit` JSON ({ candidate, sdpMLineIndex, sdpMid }).
struct CallIceCandidate: Codable, Sendable {
    var candidate: String
    var sdpMLineIndex: Int32
    var sdpMid: String?
}

/// Deterministic, order-independent call id shared by both peers
/// (mirrors `directCallId` in the web app).
func directCallId(_ a: String, _ b: String) -> String {
    [a, b].sorted().joined(separator: ":")
}

File: android/app/src/main/java/com/wsgpolar/disband/call/CallSignal.kt
package com.wsgpolar.disband.call

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Deterministic shared call id, mirroring iOS `CallSignal.directCallId`. */
fun directCallId(a: String, b: String): String =
    listOf(a, b).sorted().joinToString(":")

/**
 * Wire-compatible with the iOS `CallSignal` (and the web `useCallManager`).
 * The entire object is the Realtime broadcast payload on event `"call"`.
 *
 * `type` ∈ ring | accept | reject | cancel | handled | leave | offer | answer | ice
 */
@Serializable
data class CallSignal(
    val type: String,
    val from: String,
    val to: String? = null,
    @SerialName("callId") val callId: String? = null,
    @SerialName("callerName") val callerName: String? = null,
    @SerialName("rejecterName") val rejecterName: String? = null,
    val sdp: CallSdp? = null,
    val candidate: CallIceCandidate? = null,
)

/** Mirrors `RTCSessionDescriptionInit` JSON ({ type, sdp }). */
@Serializable
data class CallSdp(
    val type: String,
    val sdp: String,
)

/** Mirrors `RTCIceCandidateInit` JSON. */
@Serializable
data class CallIceCandidate(
    val candidate: String,
    @SerialName("sdpMLineIndex") val sdpMLineIndex: Int,
    @SerialName("sdpMid") val sdpMid: String? = null,
)